pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountBitmapAggregationFunction.java

```diff
@@ -0,0 +1,409 @@
+      // For dictionary-encoded expression, convert dictionary ids to hash code of the values
+      return convertToValueBitmap(bitmap, _dictionary);
+    } else {
+      // For serialized RoaringBitmap and non-dictionary-encoded expression, directly return the bitmap
+      return bitmap;
+    }
+  }
+
+  @Override
+  public RoaringBitmap extractGroupByResult(GroupByResultHolder groupByResultHolder, int groupKey) {
+    RoaringBitmap bitmap = groupByResultHolder.getResult(groupKey);
+    if (bitmap == null) {
+      return new RoaringBitmap();
+    }
+
+    if (_dictionary != null) {
+      // For dictionary-encoded expression, convert dictionary ids to hash code of the values
+      return convertToValueBitmap(bitmap, _dictionary);
+    } else {
+      // For serialized RoaringBitmap and non-dictionary-encoded expression, directly return the bitmap
+      return bitmap;
+    }
+  }
+
+  @Override
+  public RoaringBitmap merge(RoaringBitmap intermediateResult1, RoaringBitmap intermediateResult2) {
+    intermediateResult1.or(intermediateResult2);
+    return intermediateResult1;
+  }
+
+  @Override
+  public boolean isIntermediateResultComparable() {
+    return false;
+  }
+
+  @Override
+  public ColumnDataType getIntermediateResultColumnType() {
+    return ColumnDataType.OBJECT;
+  }
+
+  @Override
+  public ColumnDataType getFinalResultColumnType() {
+    return ColumnDataType.INT;
+  }
+
+  @Override
+  public Integer extractFinalResult(RoaringBitmap intermediateResult) {
+    return intermediateResult.getCardinality();
+  }
+
+  /**
+   * Returns the bitmap from the result holder or creates a new one if it does not exist.
+   */
+  protected static RoaringBitmap getBitmap(AggregationResultHolder aggregationResultHolder) {
+    RoaringBitmap bitmap = aggregationResultHolder.getResult();
+    if (bitmap == null) {
+      bitmap = new RoaringBitmap();
+      aggregationResultHolder.setValue(bitmap);
+    }
+    return bitmap;
+  }
+
+  /**
+   * Returns the bitmap for the given group key or creates a new one if it does not exist.
+   */
+  protected static RoaringBitmap getBitmap(GroupByResultHolder groupByResultHolder, int groupKey) {
+    RoaringBitmap bitmap = groupByResultHolder.getResult(groupKey);
+    if (bitmap == null) {
+      bitmap = new RoaringBitmap();
+      groupByResultHolder.setValueForKey(groupKey, bitmap);
+    }
+    return bitmap;
+  }
+
+  /**
+   * Helper method to set value for the given group keys into the result holder.
+   */
+  private void setValueForGroupKeys(GroupByResultHolder groupByResultHolder, int[] groupKeys, int value) {
+    for (int groupKey : groupKeys) {
+      getBitmap(groupByResultHolder, groupKey).add(value);
+    }
+  }
+
+  /**
+   * Helper method to read dictionary and convert dictionary ids to hash code of the values for dictionary-encoded
+   * expression.
+   */
+  private static RoaringBitmap convertToValueBitmap(RoaringBitmap dictIdBitmap, Dictionary dictionary) {
```

Comment:
This doesn't really look like a good use of bitmap for space efficiency purpose. We are storing the hashcodes of real values in the bitmap for result extraction
Since the range of hashcodes could be arbitrary (as opposed to 0 .. N-1 for dictionaryIds), I wonder if bitmap will end up consuming more space here?

Reply:
That is exactly the purpose of using RoaringBitmap, which is the well compressed version of the BitSet. The serialized RoaringBitmap size is much smaller than the serialized IntOpenHashSet.